File: ios/Sources/OpenUrlPlugin.swift
import SwiftRs
import Tauri
import UIKit
import WebKit


class PingArgs: Decodable {
  let value: String?
}

class OpenUrlArgs: Decodable {
  let url: String
}

class OpenUrlPlugin: Plugin {
  @objc public func openUrl(_ invoke: Invoke) throws {
    let args = try invoke.parseArgs(OpenUrlArgs.self)
    if let url = URL(string: args.url) {
      UIApplication.shared.open(url)
      invoke.resolve(nil)
    } else {
      invoke.reject("Invalid URL")
    }
}

@_cdecl("init_plugin_openurl")
func initPlugin() -> Plugin {
  return OpenUrlPlugin()
}

File: android/src/main/java/OpenUrlPlugin.kt
package com.giripriyadarshan.openurl

import android.app.Activity
import android.content.Intent
import android.webkit.WebView
import android.net.Uri
import androidx.activity.result.ActivityResult
import app.tauri.annotation.ActivityCallback
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.Plugin
import app.tauri.plugin.JSArray
import app.tauri.plugin.JSObject
import org.json.JSONException
import java.util.Collections
import java.util.concurrent.ExecutionException


@InvokeArg
internal class OpenUrlArgs {
  lateinit var Url: String
}

@TauriPlugin
class OpenUrlPlugin(private val activity: Activity) : Plugin(activity) {

    @Command
    fun openUrl(invoke: Invoke) {
        val url = invoke.parseArgs(OpenUrlArgs::class.java)
        val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(url.Url))

        startActivityForResult(invoke, browserIntent, "openUrlResult")
    }

    @ActivityCallback
    private fun openUrlResult(invoke: Invoke, result: ActivityResult) {
        invoke.resolve()
    }
}
